Fixes merge conflict with account creation transaction.

nasthagiri opened the pull request:
Master is currently broken because of transaction logic introduced by the following 2 PRs:
https://github.com/edx/edx-platform/pull/7353 (Mobile registration with Google/FB)
https://github.com/edx/edx-platform/pull/7327 (Implement full preferences API)

Broken tests on master: https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/2296/#showFailuresLink

nasthagiri commented on common/djangoapps/student/views.py:
```diff
@@ -1502,6 +1501,9 @@ def create_account_with_params(request, params):
                 raise ValidationError({'access_token': [error_message]})
 
+    # Perform operations that are non-critical parts of account creation
+    preferences_api.set_user_preference(user, LANGUAGE_KEY, get_language())
```
Note to reviewers: this call needs to be made outside of the transaction block above since it has its own `@transaction.commit_on_success` decorator.  So it doesn't rollback the outer transaction if the inner transaction succeeds.  This was an issue that was introduced by the original call to `set_user_preference`, but there was no unit test at the time to verify the transaction rollback.

With django 1.6, we'll be able to use `auto_commit`, which makes nested transactions much easier to deal with.

nasthagiri commented:
@andy-armstrong please review.
@bradenmacdonald thanks for pointing out the issue.  would you be able to review as well?

benpatterson commented:
:+1:

bradenmacdonald commented:
:+1:

cahrens commented:
:+1: But @andy-armstrong, we should audit the places calling the methods that we added the transaction decorator to. We certainly didn't consider nested calls.

cahrens commented:
And thanks for taking care of this, @nasthagiri. Hope you are having a great time in San Francisco!

andy-armstrong commented:
@nasthagiri Thanks for much for fixing this. A belated :+1: from me.

@cahrens It is a good idea to do an audit. Hopefully we have good enough test coverage that there won't be any other issues. This failure only happened because the mobile PR merged without a recent enough rebase on master, so it was passing on Jenkins on the local branch without our changes from earlier in the week.

nasthagiri commented:
@andy-armstrong @cahrens The issue existed even before the mobile PR.  The reason it was found with the mobile PR is because I added unit tests to explicitly test the rollback of a failed account creation. 

We don't have enough test coverage on account creation, unfortunately, at the moment.  
Account creation has missing converge for: transaction rollback (except for what I added), analytics, and integration tests on the logistration workflow.

cahrens commented:
Right, @andy-armstrong unit tests will only find issues if they are specifically testing for rollback behavior. We should make sure that calls that we made to our API are not in nested transaction control blocks.

We should also examine if we have been too aggressive with the decorator. For instance, why would setting a single preference need rollback behavior? I can see it for setting multiple preferences or account fields, but I don't think it makes sense for a single preference.

andy-armstrong commented:
@nasthagiri @cahrens Thanks for the explanation, that all makes sense to me now. I had no idea that you couldn't nest calls using this decorator, and that you had to explicitly test for rollbacks. This Django transaction support seems like a very brittle solution for something that is so critical. 

Until we move up to 1.6, it seems that we shouldn't decorate any of the APIs because the goal is that they can be used in a variety of contexts. Maybe the transaction logic should be moved into the outermost code (the views themselves, I assume) so that we don't have any risk of nested transactions. Can we put the decorators onto the views or would we need to hand-code the logic?
